fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,841 @@
+    predictor using an API that is similar to estimators in `sklearn`.
+
+    As per [1]_, the neural network models :code:`predictor_model` and
+    :code:`adversary_model` may not have a discrete prediction at the end of the
+    model, as gradients do not propagate through a discretization step.
+    Instead, even if we desire discrete predictions in the end, we must
+    output the sigmoidal or soft-max during training. You can supply the final
+    discrete prediction function later,
+    see parameter :code:`prediction_function`.
+
+    The distribution types of :code:`y` and :code:`sensitive_features`
+    are set by their preprocessor :code:`y_transform` and :code:`a_transform`
+    respectively. The default transformer :code:`FloatTransformer("auto")`
+    attempts to automatically infer
+    whether to assume binomial, multinomial, or normally distributed data.
+    You can force the transformer to assume one of the above by passing
+    :code:`FloatTransformer("binary")`, :code:`FloatTransformer("category")`,
+    :code:`FloatTransformer("continuous")` instead.
+    Loss functions and decision functions
+    for the target and sensitive features are also predefined for keywords
+    "binary", "category", and "continuous"
+    For more information, visit the user guide.
+
+    Parameters
+    ----------
+    backend : str, BackendEngine, default = 'auto'
+        The backend to use. Must be one of :code:`'torch'`, :code:`'tensorflow'`,
+        or :code:`'auto'` which indicates PyTorch, TensorFlow, or to
+        automatically infer
+        the backend from the :code:`predictor_model` and which are installed.
+        You can also pass in a BackendEngine class.
+
+    predictor_model : list, torch.nn.Module, tensorflow.keras.Model
+        The predictor model to train.
+        Instead of a neural network model, we can pass a list of keywords
+        :math:`k_1, k_2, \dots` that indicate either
+        the number of nodes :math:`k_i` (if :math:`k_i` is integer) or a keyword
+        that indicates an activation function (if :math:`k_i` is a string) or
+        a layer or activation function instance directly (if :math:`k_i` is
+        callable). The default parameter is :code:`[]`, which indicates
+        a neural network without any hidden layers.
+        However, the number of nodes in the input
+        and output layer are automatically inferred from data, and the final
+        activation function (such as softmax for categorical
+        predictors) are inferred from data.
+        If :code:`backend` is specified, you cannot pass a model
+        that uses a different backend.
+
+    adversary_model : list, torch.nn.Module, tensorflow.keras.Model
+        The adversary model to train. Defined similarly as predictor_model.
+        Must be the same type as the
+        :code:`predictor_model`.
+
+    predictor_loss : str, callable, default = 'auto'
```

Comment:
I don't love the current keyword choices for `predictor_loss`, `adversary_loss`, because they seem to refer to the type of the target, rather than the loss. If we go that route, we should consider using something similar to sklearn's [target types](https://scikit-learn.org/stable/modules/generated/sklearn.utils.multiclass.type_of_target.html#sklearn.utils.multiclass.type_of_target).

Alternatively, we could use keywords that describe the loss, like `"square_loss"`, `"logistic_loss"`... but let me think a bit more about this.

Reply:
You raise an excellent point. Ideally, we'd also provide something like `Y_distribution_type` and `A_distribution_type`. I've thought about this before, but I can't remember why I let go of the idea. We should not get rid of the loss parameters though, and we would have to be sure that the inferred distribution type of the preprocessor agrees.

Reply:
I had some further thoughts on this.

I think that the very basic question is whether we want to represent binary classification problems via networks with a single output or two outputs... so that should be decided upfront.

With that said, what do you think about the following tweaks of the current API:
* base class (currently `AdversarialFairness`)
   * allow specifying `y_transform`, `sf_transform`, `predictor_loss`, `adversary_loss`, `predictor_function`
   * they can take values `None` (no transformation; this might still mean casting ints as a floats?), `'auto'` (default), a callable, and additionally:
     * for `predictor/adversary_loss`, we support `'logistic_loss'`, `'square_loss'`
       * if needed, we could also distinguish `'multinomial_logistic_loss'` (which acts on one-hot encoding), but this could be inferred from the number of outputs of the network
     * for `y/sf_transform`, we support `'one_hot_encoder'`
     * for `predictor_function`, we support `'argmax'` (for one-hot representation) and `'threshold'` (for one-output representation of binary classifiers), with an additional argument `threshold_value` with the default value 0
  
* `AdversarialFairnessClassifier`
  * allow specifying `sf_transform`, `adversary_loss`
  * fill in `y_transform`, `predictor_loss`, `adversary_loss`, `predictor_function` so as to support (multinomial) logistic regression; I wouldn't allow overriding these (if that's what you want to do, you can just call the base class)
  * besides `predict` and `decision_function`, we should consider supporting `predict_proba` and `predict_log_proba`

* `AdversarialFairnessRegressor`
  * allow specifying `sf_transform`, `predictor_loss`, `adversary_loss`
  * fill in `y_transform`, `predictor_function` as `None`

And one more idea--not sure how much I like it, but something that occurred to me. We could change:
* `predictor_loss` -> `y_loss`
* `adversary_loss` -> `sf_loss`

Reply:
I like your thoughts! I really appreciate you taking the time! 

I just want to reiterate that we (it was a suggestion from adrin or hilde I think) made the design choice that as much is automatically inferred from data as possible, so that means what kind of transform/loss/predictor function to use. Looking at this now, I am not so sure why I included `y/sf_transform` as parameters, because without this `FloatTransformer` class thingy (`FloatTransform` is the default `y/sf_transform`) we lose this automatic inference. The nice thing is that if the user has already applied a one-hot-encoding, the FloatTransform class will still infer and tell AdversarialFairness that the data is categorical, which is required to infer what loss function to use. I agree that the current API needs changing, but I do not see yet how we can resolve this nicely.

> * for `y/sf_transform`, we support `'one_hot_encoder'`

I think I like this keyword-style better than how it is currently done, but I'm still on the fence. Currently, you can achieve this `sf_transform='one_hot_encoder'` with `sf_transform=FloatEncoder('categorical')`, so what you suggest is definitely cleaner. What do you think should happen if the user provides a custom transform? Should we then require the user to (1) also pass loss functions, because we can no longer infer them from FloatEncoder? Or (2) should we do the inferring outside of FloatEncoder? or (3) should we not even let the user pass a custom transform? I find this a difficult choice because all options feel bad in some sense. I've actually switched implementations from (2) to (1) in the past, but I am now tending back towards (2) and using the keywords you suggested. I'd love to hear your thoughts on this. @adrinjalali you were quite involved in this design choice regarding preprocessing in the past, so perhaps you can help us here. 

> * for `predictor/adversary_loss`, we support `'logistic_loss'`, `'square_loss'`

Currently we accept `'binary'`, `'category'`, `'continuous'`, I chose those because they are descriptive of the distribution that you assume, but I understand that you favor more precise names such as `'logistic_loss'`, `'square_loss'`, or `'argmax'`? I might agree actually.

> * fill in `y_transform`, `predictor_loss`, `adversary_loss`, `predictor_function` so as to support (multinomial) logistic regression;

Then we'd still need to infer whether it is categorical or binary I'd say, hence I'd still love to know the distribution type somehow (rather through inferring it from the data than through a keyword parameter)

> And one more idea--not sure how much I like it, but something that occurred to me.

I like this!


> * besides `predict` and `decision_function`, we should consider supporting `predict_proba` and `predict_log_proba`

Yeah sure!

Reply:
Let me try to answer your questions--but also let me know if I left something unanswered!

**What to do about `'auto'` loss functions and `predictor_function` with custom `y/sf_transform`.**
* The behavior should be the same as what we do when the transform is `None`. I'm not sure what you do currently, but a sensible option would be an automatic inference among:
  1. univariate {0.0,1.0} (-> univariate logistic model with sign-based prediction function),
  2. one-hot-encoded categorical (-> multinomial logistic with argmax prediction function),
  3. continuous univariate or vectors (-> square loss/L2 norm with identity prediction function),
  4. otherwise: throw an exception

**Loss versus link function ("distribution assumption").**
* I think by "distribution assumption" you mean the [link function](https://en.wikipedia.org/wiki/Generalized_linear_model#Link_function) that is used in generalized linear models. I'd be in favor of keeping the idea of link (which is basically our `prediction_function`) separate from loss function. For example, it would be awkward to specify Huber loss (for robust regression) using the language of link functions.

**Categorical vs. binary encoding in classification.**
* For this one, I'm not sure that we have any other option other than inferring from the provided `y`. My idea would be to support the binary vs. multiclass classification similarly to sklearn's [LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).

Reply:
Thanks, this was really clarifying!

> What to do about `'auto'` loss functions and `predictor_function` with custom `y/sf_transform`.

Oh so doing this inference if the transform is None, and otherwise just do the transform and infer then. That sounds good. Basically decoupling this inference from the transform. Okay! Apart from that, I make the same inferences (the i. through iv.). (I initially chose to be a bit more general and accept mixed column types (so for instance binary and continuous columns), but we decided there was no use for this.)

> Loss versus link function ("distribution assumption").

I must say I am learning a lot here, I was not familiar with these terms. I understand a reason for keeping these things seperate now, thanks!

Reply:
@MiroDudik Hey, I finally got around to this, was a bit more work than I anticipated. What do you think? Internally, I still use "binary", "category", "continuous", because I feel this may be relevant to know later, but the API changed and accepts the better more concrete terms now.

As for the `predict_(log_)proba`, I am not sure how to interpret this. In case of a single output neuron $\hat y$, should we unfold it into a vector with separate probabilities for $\hat y = 0$ and $\hat y = 1$? And what if there are more weird cases? For now, we do have `decision_function`, which might give enough flexibility?

Reply:
See my comment above -> if we specialize to logistic models, then `predict_proba` and `predict_log_proba` are obvious, right? These would be only provided for AdversarialMitigationClassifier.